### DISCUSSION TITLE
[Louvain Clustering] Different resolutions leading to same number of clusters but different labels

### DISCUSSION THREAD

**Discussion by prakritipaul** [sknetwork-team/scikit-network#537] [Q&A] [answered]
Hi! 

I have been using Louvain Clustering and noticed that a different resolutions, I may get the same number of clusters, but different cluster labels. 

In the example below (which I got after clustering)…
RES_DICT = {Decimal('0.4'): (2, 1.0, 0, 282),
 			Decimal('0.5'): (2, 1.0, 0, 269),
			Decimal('0.8'): (2, 1.0, 0, 236),
			 Decimal('0.9'): (2, 1.0, 1, 225),
			 Decimal('1'): (3, 1.0, 0, 162),
			 Decimal('1.1'): (3, 1.0, 0, 161),
			 Decimal('1.2'): (3, 1.0, 1, 157),
			 Decimal('1.3'): (3, 1.0, 1, 157),
			 Decimal('1.4'): (3, 1.0, 1, 151)}

The keys of this dictionary RES_DICT are resolution values.
The values are tuples with (number of clusters, a “disagreement ratio”, true/first cluster label, and a cutoff index)
You can ignore the disagreement ratio here.

If there are two clusters, my cluster labels can either be  0 or 1.
If there are 3 clusters, my cluster labels can be 0, 1, or 2.

Taking 2 clusters as an example [0, 0, 0, 1, 1, 1], if 0 is the true cluster label, the cutoff index is the index at which the cluster label is no longer 0 (and is now 1). Thus, in this case, the cutoff index is 3 (0-indexed). The cutoff index is also the size of the first cluster.

If I have 3 clusters [0, 0, 1, 2, 2, 2] and 0 is the true cluster label, the cutoff index is 2.

Thus, as you can observe in this dictionary, for different resolutions, I can have the same number of clusters but different cutoffs, which implies that labels/membership of points is different. 

e.g. 
Decimal('0.4'): (2, 1.0, 0, 282),
Decimal('0.5'): (2, 1.0, 0, 269),
Decimal('0.8'): (2, 1.0, 0, 236),
Decimal('0.9'): (2, 1.0, 1, 225)

With resolutions 0.4, 0.5, 0.8, 0.9, I get 2 clusters from all, but the cutoff indices are 282, 269, 236, and 225 respectively! 

Why is this the case?

Thanks!
PC

**Comment by tbonald** [comment, marked as answer]
Thanks for your feedback.
This is the case because the objective function that Louvain maximizes is different. Increasing the resolution puts more weight on the "diversity term" (second term of modularity), meaning that you either increase the number of clusters or make them more balanced.

**Reply by prakritipaul** [reply]
Thank you, that is very clear!
